core/src/main/scala/org/apache/spark/status/InMemoryStoreCheckpoint.scala

```diff
@@ -0,0 +1,136 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.status
+
+import java.io.BufferedOutputStream
+import java.util.concurrent.{CountDownLatch, TimeUnit}
+
+import org.apache.hadoop.fs.{FileUtil, Path}
+
+import org.apache.spark.SparkConf
+import org.apache.spark.deploy.SparkHadoopUtil
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config._
+import org.apache.spark.internal.config.Status._
+import org.apache.spark.scheduler.EventLoggingListener
+import org.apache.spark.serializer.JavaSerializer
+import org.apache.spark.status.api.v1
+import org.apache.spark.util.{ThreadUtils, Utils}
+import org.apache.spark.util.kvstore.InMemoryStore
+
+case class InMemoryStoreSnapshot(
+    store: InMemoryStore,
+    eventsNum: Int,
+    finished: Boolean) extends Serializable
+
+
+private[spark] class InMemoryStoreCheckpoint(
+    store: InMemoryStore,
+    conf: SparkConf,
+    listener: AppStatusListener) extends Logging {
+  var lastRecordEventsNum: Int = 0
+  var finished: Boolean = false
+
+  // used to count the number of processed events in a live AppStatusListener
+  private var processedEventsNum = 0
+  private val batchSize = conf.get(IMS_CHECKPOINT_BATCH_SIZE)
+  private val bufferSize = conf.get(IMS_CHECKPOINT_BUFFER_SIZE).toInt
+  private var latch = new CountDownLatch(0)
+  @volatile var isDone = true
+  // a JavaSerializer used to serialize InMemoryStoreSnapshot
+  private val serializer = new JavaSerializer(conf).newInstance()
+  private val logBaseDir = Utils.resolveURI(conf.get(EVENT_LOG_DIR).stripSuffix("/"))
+  private lazy val ckpPath = getCheckpointPath
+  private val hadoopConf = SparkHadoopUtil.get.newConfiguration(conf)
+  private val fileSystem = Utils.getHadoopFileSystem(logBaseDir, hadoopConf)
+  private val executor = ThreadUtils.newDaemonSingleThreadExecutor(
+    "inmemorystore-checkpoint-thread")
+  // should be inited before the first time checkpoint
+  var appInfo: v1.ApplicationInfo = _
+
+  private val checkpointTask = new Runnable {
+    override def run(): Unit = Utils.tryLogNonFatalError(doCheckpoint())
+  }
+
+  private def getCheckpointPath: String = {
+    val appId = appInfo.id
+    val appAttemptId = appInfo.attempts.head.attemptId
+    EventLoggingListener.getLogPath(logBaseDir, appId, appAttemptId, None)
+  }
+
+  def await(): Unit = latch.await()
+
+  def eventInc(finish: Boolean = false): Unit = {
+    processedEventsNum += 1
+    val shouldCheckpoint = !finished && (processedEventsNum - lastRecordEventsNum >=
+      batchSize || finish)
+    if (shouldCheckpoint) {
+      // flush to make sure that all processed events' related data have write into InMemoryStore
+      listener.flush(listener.update(_, System.nanoTime()))
+      latch = new CountDownLatch(1)
```

Comment:
Is it thread-safe to create a new `CountDownLatch` here?

Reply:
Since all events are processed in the single dispatchThread, so we'll only have that one thread to create new `CountDownLatch`. But I just realize that variable `latch` maybe override if the number of backlog events exceeds checkpoint batch size while last checkpoint is still in progress. So, I update the `shouldCheckpoint` condition to `isDone && (finish || processedEventsNum - lastRecordEventsNum >= batchSize)` to cover that corner case.